src/transformStateWithClones.js

```diff
@@ -1,13 +1,37 @@
 'use strict';
-
 /**
  * @param {Object} state
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+
+function transformStateWithClones(initialState, actions) {
+  const stateHistory = [];
+  let state = { ...initialState };
+
+  for (const action of actions) {
+    let newState;
+
+    if (action.type === 'clear') {
+      newState = {};
+    } else {
+      newState = { ...state };
+
+      if (action.type === 'addProperties') {
+        Object.assign(newState, action.extraData);
+      } else if (action.type === 'removeProperties') {
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+      }
+    }
+
+    stateHistory.push(newState);
+    state = newState;
```

Comment:
[CODE STYLE]: Instead of creating a new variable 'newState' in each iteration, you can simply use 'state' to apply the changes and then push the copy to 'stateHistory' at the end of each loop cycle. This adheres to the DRY principle and ensures that you perform the action of copying the state just once after all modifications are done.